### ISSUE TITLE
How to infer dummy Input for partial module export with torch.export

### ISSUE THREAD

**Issue by mhs4670go** [Samsung/TICO#63]
When exporting only a part of your PyTorch model (like a single layer) with `torch.export`, it’s not always clear what the dummy input should be—especially for complex submodules with custom forward signatures.

### Typical approaches

Register hooks on the module to capture input/output shapes during a real forward pass. However, both approaches have caveats. Hooks can capture tensor shapes, but it’s hard to recover keyword argument names and default values.

### Recommended approach

Temporarily patch the submodule’s forward method using `types.MethodType` to capture both positional and keyword arguments, including shapes and names.

#### Patch the Submodule’s Forward Method

```python
import torch
import types

def record_inputs(self, *args, **kwargs):
    # Save both args and kwargs for later use
    record_inputs.args = args
    record_inputs.kwargs = kwargs
    # Optionally, call the real forward for correctness
    return self._real_forward(*args, **kwargs)

# Suppose you want to export model.layer2
submodule = model.layer2

# Save original forward
submodule._real_forward = submodule.forward

# Properly bind the recorder method to submodule (preserves 'self')
submodule.forward = types.MethodType(record_inputs, submodule)

# Run a normal forward pass with realistic input to the full model
_ = model(some_real_input)

# Now retrieve the captured input for the submodule
args = record_inputs.args
kwargs = record_inputs.kwargs
print("Captured dummy input for submodule:")
print("args:", [a.shape if isinstance(a, torch.Tensor) else type(a) for a in args])
print("kwargs:", {k: (v.shape if isinstance(v, torch.Tensor) else type(v)) for k, v in kwargs.items()})

# Restore the original forward (important!)
submodule.forward = submodule._real_forward
del submodule._real_forward

# Now use (args, kwargs) as the dummy input for torch.export
# Example:
# torch.export(submodule, args, kwargs, ...)
```

### Why Use `types.MethodType` instead of just overriding the forward method?

Ensures the patched forward method receives the correct self argument. Avoids subtle bugs, especially for submodules where forward accesses self attributes or methods. More robust in complex inheritance or when submodule’s forward is nontrivial.

By temporarily replacing the submodule’s forward with a types.MethodType-wrapped recorder, you can safely and reliably capture all input arguments for dummy input construction—making partial module export with torch.export straightforward.

**Comment by jinevening** [comment]
What is the advantage of using `types.Method` compared to `register_foward_hook` with `with_kwargs = True`?

**Comment by mhs4670go** [comment]
@jinevening Ah, I didn't know there was `with_kwargs` option! I revise above comments. Thanks.

**Comment by mhs4670go** [comment]
@zhxchen17 

Thanks for the comment—I totally agree that we should push toward a single or user-friendly export call. For example, we could simply extend `torch.export.export` to accept a target option:

```python
m = load_model()
ep = torch.export.export(
    m,
    model_inputs,
    target=m.target_module,  # specify which submodule to export
)
```


Or, here are a few alternative interface designs you might consider alongside it later.

### 1. Decorator-style export targets

Mark submodules at definition time, then call `export(model, inputs)` without extra args:

```python
class MyModel(nn.Module):
    @torch.export.exportable    # or @capture_for_export / @export_target
    def target_module(self, x):
        return ...

model = MyModel()
ep = torch.export.export(model, model_inputs)
```

Pros

- Call site is as simple as `export(model, inputs)`
- Familiar to anyone who’s used decorator APIs like `torch.compile()`

Cons

- Requires decorating modules up-front
- Hides the “capture” mechanism inside the decorator

### 2. Context-manager capture

Scope your capture in a with block, then export using the captured context:

```python
with torch.export.capture_targets(model, ["target_module"]) as ctx:
    _ = model(model_inputs)

ep = torch.export.export(model, sample_ctx=ctx)
```

Pros

- Very explicit about when and what you’re capturing
- Easy to scope to multiple targets or runs

Cons

- Introduces an extra with block

### 3. Configuration object

Bundle all export options into a reusable config:

```python
cfg = torch.export.ExportConfig(
    targets=["features.conv1", "classifier"],
    capture_first_n=5
)

ep = torch.export.export(model, model_inputs, config=cfg)
```

Pros

- Keeps the export signature clean even as options grow
- Configs can be saved, shared, or versioned

Cons

- Requires learning a new `ExportConfig` class

### 4. Builder/DSL style

Chain methods for maximum readability:

```python
ep = (
    torch.export
      .for_model(model)
      .with_inputs(model_inputs)
      .targeting("target_module")   # or .targets([...])
      .run(output_path="model_export.pt")
)
```

Pros

- Self-documenting, discoverable API
- Easy to extend with new steps/options

Cons

- More ceremony and a slightly higher learning curve

**Comment by glistening** [comment]
I read this issue again. It looks different issue.

> When exporting only a part of your PyTorch model (like a single layer) with torch.export

I don't want to export only a part, but the whole module.

But the top module (for my case, `LlamaForCausalLM`) itself is used for both prefill and decode.

`generate` prepares the appropriate inputs and control the logic for prefill and decode.

In this sense, top module may be considered as submodule.

I hope the solution for this issue solve my problem also. :)
